src/runtime_src/core/pcie/driver/linux/xocl/subdev/ps.c

```diff
@@ -0,0 +1,202 @@
+/*
+ * A GEM style device manager for PCIe based OpenCL accelerators.
+ *
+ * Copyright (C) 2016-2018 Xilinx, Inc. All rights reserved.
+ *
+ * Authors: Min.Ma@xilinx.com
+ *
+ * This software is licensed under the terms of the GNU General Public
+ * License version 2, as published by the Free Software Foundation, and
+ * may be copied, distributed, and modified under those terms.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the
+ * GNU General Public License for more details.
+ */
+
+#include "../xocl_drv.h"
+#include "mgmt-ioctl.h"
+
+#define MAX_RETRY       50
+#define RETRY_INTERVAL  100       //ms
+
+#define RESET_REG	0xC
+
+#define ERT_READY_MASK  0x8
+#define RES_DONE_MASK   0x4
+#define RES_TYPE_MASK   0x3
+
+#define SK_RESET	0x1
+#define ERT_RESET	0x2
+#define SYS_RESET	0x3
+
+#define READ_REG32(ps, off)	\
+	XOCL_READ_REG32(ps->base_addr + off)
+#define WRITE_REG32(ps, val, off)       \
+	XOCL_WRITE_REG32(val, ps->base_addr + off)
+
+struct xocl_ps {
+	struct platform_device	*pdev;
+	void __iomem		*base_addr;
+};
+
+/*
+ * Processor system reset supports 3 type of reset
+ * Host set reset type in the scratchpad register then wait PS
+ * set reset done bit.
+ */
+static void ps_reset(struct platform_device *pdev, int type)
+{
+	struct xocl_ps *ps;
+	u32 reg;
+	int retry = 0;
+
+	xocl_info(&pdev->dev, "Reset Processor System...");
+	ps = platform_get_drvdata(pdev);
+	if (!ps)
+		return;
+
+	reg = READ_REG32(ps, RESET_REG);
+	/* Set reset type in scratchpad register */
+	switch(type) {
+	case 1:
+		xocl_info(&pdev->dev, "Soft Kernel reset...");
+		reg = (reg & ~RES_TYPE_MASK) | SK_RESET;
+		break;
+	case 2:
+		xocl_info(&pdev->dev, "ERT reset...");
+		reg = (reg & ~RES_TYPE_MASK) | ERT_RESET;
+		break;
+	case 3:
+		xocl_info(&pdev->dev, "SYS reset...");
+		reg = (reg & ~RES_TYPE_MASK) | SYS_RESET;
+		break;
+	default:
+		xocl_info(&pdev->dev, "Unknown reset type");
+	}
+	WRITE_REG32(ps, reg, RESET_REG);
+
+	do {
+		reg = READ_REG32(ps, RESET_REG);
+		msleep(RETRY_INTERVAL);
+	} while (retry++ < MAX_RETRY && !(reg & RES_DONE_MASK));
+
+	if (retry >= MAX_RETRY) {
+		xocl_err(&pdev->dev, "Reset time out");
+		return;
+	}
+
+	/* Clear reset done bit */
+	reg &= ~RES_DONE_MASK;
+	WRITE_REG32(ps, reg, RESET_REG);
+}
+
+/* Wait Processor system goes into ready status */
+static void ps_wait(struct platform_device *pdev)
+{
+	struct xocl_ps *ps;
+	u32 reg;
+	int retry = 0;
+
+	xocl_info(&pdev->dev, "Wait Processor System ready...");
+	ps = platform_get_drvdata(pdev);
+	if (!ps)
+		return;
+
+	/*
+	 * The PS need about 30 ~ 40 seconds for reboot
+	 * 50 seconds are more than enough
+	 */
+	msleep(50000);
```

Comment:
why not poll the status every, say, 5 seconds, so that you might finish waiting earlier, if the reboot is done for < 50 seconds?

Reply:
Done.